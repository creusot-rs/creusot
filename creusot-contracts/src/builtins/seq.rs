use creusot_contracts_proc::*;

use crate::Int;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T>(std::marker::PhantomData<T>);

impl<T> Seq<T> {
    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index(self, _: Int) -> T {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }
}
