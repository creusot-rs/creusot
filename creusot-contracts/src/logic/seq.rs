use creusot_contracts_proc::*;

use crate::logic::*;
use crate::Int;
use std::ops::Index;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T: ?Sized>(std::marker::PhantomData<T>);

impl<T> EqLogic for Seq<T> {
    #[predicate]
    #[creusot::builtins = "seq.Seq.(==)"]
    fn log_eq(self, _: Self) -> bool {
        true
    }

    #[predicate]
    fn log_ne(self, o: Self) -> bool {
        pearlite! { !(self == o)}
    }

    #[logic]
    fn eq_ne(_: Self, _: Self) {}

    #[logic]
    fn refl(_: Self) {}

    #[logic]
    fn symmetry(_: Self, _: Self) {}

    #[logic]
    fn transitivity(_: Self, _: Self, _: Self) {}
}

impl<T> Seq<T> {
    #[logic]
    pub fn get(self, ix: Int) -> Option<T> {
        if ix < self.len() {
            Some(*self.index(ix))
        } else {
            None
        }
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq_ext.SeqExt.subsequence"]
    pub fn subsequence(self, _: Int, _: Int) -> Self {
        std::process::abort()
    }

    #[logic]
    pub fn tail(self) -> Self {
        self.subsequence(1, self.len())
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }

    #[predicate]
    pub fn permutation_of(self, o: Self) -> bool {
        self.permut(o, 0, self.len())
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.permut"]
    pub fn permut(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.exchange"]
    pub fn exchange(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }
}

// A hack which allows us to use [..] notation for sequences.
// Relies on the fact we don't enforce that implementations of traits are of
// the same function type as the trait signature.. When this is addressed
// the following instance will error.
impl<T> std::ops::Index<Int> for Seq<T> {
    type Output = T;

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.get"]
    fn index(&self, _: Int) -> &T {
        std::process::abort()
    }
}
