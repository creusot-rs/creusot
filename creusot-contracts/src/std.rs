pub mod clone;
pub mod eq;
pub mod mem;
pub mod ord;
pub mod vec;

mod slice;

pub use clone::*;
pub use eq::*;
pub use ord::*;
pub use vec::*;
