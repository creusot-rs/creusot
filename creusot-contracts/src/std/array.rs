use crate::{invariant::*, *};

impl<T, const N: usize> Invariant for [T; N] {
    #[predicate(prophetic)]
    #[open]
    #[creusot::trusted_ignore_structural_inv]
    fn invariant(self) -> bool {
        pearlite! { inv(self@) }
    }
}

impl<T, const N: usize> ShallowModel for [T; N] {
    type ShallowModelTy = Seq<T>;

    #[logic]
    #[trusted]
    #[open]
    #[creusot::builtins = "prelude.prelude.Slice.id"]
    // TODO:
    // #[ensures(result.len() == N@)]
    // Warning: #[ensures] and #[trusted] are incompatible, so this might require
    fn shallow_model(self) -> Self::ShallowModelTy {
        pearlite! { absurd }
    }
}

impl<T: DeepModel, const N: usize> DeepModel for [T; N] {
    type DeepModelTy = Seq<T::DeepModelTy>;

    #[logic]
    #[trusted]
    #[open(self)]
    // TODO
    // #[ensures(result.len() == N@)]
    #[ensures(self.shallow_model().len() == result.len())]
    #[ensures(forall<i: _> 0 <= i && i < result.len() ==> result[i] == self[i].deep_model())]
    fn deep_model(self) -> Self::DeepModelTy {
        pearlite! { absurd }
    }
}
