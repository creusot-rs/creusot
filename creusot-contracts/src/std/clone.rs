use crate::*;
pub use ::std::clone::*;

#[cfg(creusot)]
pub use creusot_contracts_proc::Clone;

extern_spec! {
    mod std {
        mod clone {
            trait Clone {
                // TODO:
                // Requiring result == *self seems too strong since objects can
                // contain information which is lost by the clone (e.g., capacity of
                // vectors...).
                #[ensures(result == *self)]
                fn clone(&self) -> Self;
            }
        }
    }
}
