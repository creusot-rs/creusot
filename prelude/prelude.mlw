module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
