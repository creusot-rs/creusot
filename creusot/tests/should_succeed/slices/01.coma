
module C01_IndexSlice
  use prelude.UInt32
  use seq.Seq
  predicate invariant'1 (self : Seq.seq uint32) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'1 (self:Seq.seq uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'1 self} (! return' {result}) ]
    
  predicate inv'1 (_x : Seq.seq uint32)
  let rec inv'1 (_x:Seq.seq uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'1 _x} (! return' {result}) ]
    
  axiom inv'1 : forall x : Seq.seq uint32 . inv'1 x = true
  use prelude.Slice
  predicate invariant'0 (self : slice uint32) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'0 (self:slice uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'0 self} (! return' {result}) ]
    
  predicate inv'0 (_x : slice uint32)
  let rec inv'0 (_x:slice uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'0 _x} (! return' {result}) ]
    
  axiom inv'0 : forall x : slice uint32 . inv'0 x = true
  use seq.Seq
  use prelude.Slice
  use prelude.UIntSize
  use prelude.UIntSize
  use prelude.Int
  constant max'0 : usize = (18446744073709551615 : usize)
  function shallow_model'1 (self : slice uint32) : Seq.seq uint32
  let rec shallow_model'1 (self:slice uint32) (return'  (ret:Seq.seq uint32))= {[@expl:precondition] [#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'0 self}
    any [ return' (result:Seq.seq uint32)-> {result = shallow_model'1 self} (! return' {result}) ] 
  axiom shallow_model'1_spec : forall self : slice uint32 . ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'0 self)
   -> ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 4 19 50] inv'1 (shallow_model'1 self))
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 42] shallow_model'1 self = Slice.id self)
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 17 14 17 41] Seq.length (shallow_model'1 self)
  <= UIntSize.to_int (max'0 : usize))
  use prelude.Borrow
  function shallow_model'0 (self : slice uint32) : Seq.seq uint32 =
    [#"../../../../../creusot-contracts/src/model.rs" 90 8 90 31] shallow_model'1 self
  let rec shallow_model'0 (self:slice uint32) (return'  (ret:Seq.seq uint32))= any
    [ return' (result:Seq.seq uint32)-> {result = shallow_model'0 self} (! return' {result}) ]
    
  use prelude.Intrinsic
  let rec index_slice (a:slice uint32) (return'  (ret:uint32))= {[#"../01.rs" 5 11 5 24] 10
    < Seq.length (shallow_model'0 a)}
    (! bb0
    [ bb0 = s0
      [ s0 =  [ &_3 <- [#"../01.rs" 7 6 7 8] (10 : usize) ] s1
      | s1 =  [ &_4 <- Slice.length a ] s2
      | s2 = UIntSize.lt {_3} {_4} (fun (_ret':bool) ->  [ &_5 <- _ret' ] s3)
      | s3 = {[@expl:index in bounds] [#"../01.rs" 7 4 7 9] _5} s4
      | s4 = bb1 ]
      
    | bb1 = s0 [ s0 = Slice.get <uint32> {a} {_3} (fun (r'0:uint32) ->  [ &_0 <- r'0 ] s1) | s1 = return' {_0} ]  ]
    )
    [ & _0 : uint32 = any_l () : uint32
    | & a : slice uint32 = a
    | & _3 : usize = any_l () : usize
    | & _4 : usize = any_l () : usize
    | & _5 : bool = any_l () : bool ]
     [ return' (result:uint32)-> (! return' {result}) ] 
end
module C01_IndexMutSlice
  use prelude.UInt32
  use seq.Seq
  predicate invariant'1 (self : Seq.seq uint32) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'1 (self:Seq.seq uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'1 self} (! return' {result}) ]
    
  predicate inv'1 (_x : Seq.seq uint32)
  let rec inv'1 (_x:Seq.seq uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'1 _x} (! return' {result}) ]
    
  axiom inv'1 : forall x : Seq.seq uint32 . inv'1 x = true
  use prelude.Slice
  predicate invariant'0 (self : slice uint32) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 8 8 8 12] true
  let rec invariant'0 (self:slice uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'0 self} (! return' {result}) ]
    
  predicate inv'0 (_x : slice uint32)
  let rec inv'0 (_x:slice uint32) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'0 _x} (! return' {result}) ]
    
  axiom inv'0 : forall x : slice uint32 . inv'0 x = true
  use prelude.UInt32
  use seq.Seq
  use prelude.Slice
  use prelude.UIntSize
  use prelude.UIntSize
  use prelude.Int
  constant max'0 : usize = (18446744073709551615 : usize)
  use seq.Seq
  function shallow_model'1 (self : slice uint32) : Seq.seq uint32
  let rec shallow_model'1 (self:slice uint32) (return'  (ret:Seq.seq uint32))= {[@expl:precondition] [#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'0 self}
    any [ return' (result:Seq.seq uint32)-> {result = shallow_model'1 self} (! return' {result}) ] 
  axiom shallow_model'1_spec : forall self : slice uint32 . ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'0 self)
   -> ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 4 19 50] inv'1 (shallow_model'1 self))
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 42] shallow_model'1 self = Slice.id self)
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 17 14 17 41] Seq.length (shallow_model'1 self)
  <= UIntSize.to_int (max'0 : usize))
  function index_logic'0 [@inline:trivial] (self : slice uint32) (ix : int) : uint32 =
    [#"../../../../../creusot-contracts/src/logic/ops.rs" 43 8 43 31] Seq.get (shallow_model'1 self) ix
  let rec index_logic'0 (self:slice uint32) (ix:int) (return'  (ret:uint32))= any
    [ return' (result:uint32)-> {result = index_logic'0 self ix} (! return' {result}) ]
    
  use prelude.Borrow
  function shallow_model'0 (self : borrowed (slice uint32)) : Seq.seq uint32 =
    [#"../../../../../creusot-contracts/src/model.rs" 108 8 108 31] shallow_model'1 ( * self)
  let rec shallow_model'0 (self:borrowed (slice uint32)) (return'  (ret:Seq.seq uint32))= any
    [ return' (result:Seq.seq uint32)-> {result = shallow_model'0 self} (! return' {result}) ]
    
  use prelude.Intrinsic
  predicate resolve'0 (self : borrowed (slice uint32)) =
    [#"../../../../../creusot-contracts/src/resolve.rs" 26 20 26 34]  ^ self =  * self
  let rec resolve'0 (self:borrowed (slice uint32)) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = resolve'0 self} (! return' {result}) ]
    
  let rec index_mut_slice (a:borrowed (slice uint32)) (return'  (ret:()))= {[#"../01.rs" 10 11 10 24] Seq.length (shallow_model'0 a)
    = 5}
    (! bb0
    [ bb0 = s0
      [ s0 =  [ &_4 <- [#"../01.rs" 13 6 13 7] (2 : usize) ] s1
      | s1 =  [ &_5 <- Slice.length ( * a) ] s2
      | s2 = UIntSize.lt {_4} {_5} (fun (_ret':bool) ->  [ &_6 <- _ret' ] s3)
      | s3 = {[@expl:index in bounds] [#"../01.rs" 13 4 13 8] _6} s4
      | s4 = bb1 ]
      
    | bb1 = s0
      [ s0 = Slice.get <uint32> { * a} {_4}
          (fun (l'0:uint32) ->
            Slice.set <uint32> { * a} {_4} {[#"../01.rs" 13 11 13 12] (3 : uint32)}
              (fun (l'1:slice uint32) ->  [ &a <- { a with current = l'1 ; } ] s1))
      | s1 = -{resolve'0 a}- s2
      | s2 =  [ &_0 <- [#"../01.rs" 12 38 14 1] () ] s3
      | s3 = return' {_0} ]
       ]
    )
    [ & _0 : () = any_l () : ()
    | & a : borrowed (slice uint32) = a
    | & _4 : usize = any_l () : usize
    | & _5 : usize = any_l () : usize
    | & _6 : bool = any_l () : bool ]
    
    [ return' (result:())-> {[@expl:postcondition] [#"../01.rs" 11 10 11 23] UInt32.to_int (index_logic'0 ( ^ a) 2) = 3}
      (! return' {result}) ]
    
end
module Core_Option_Option_Type
  type t_option 't =
    | C_None
    | C_Some 't
    
  function any_l (_ : 'b) : 'a
  let rec none < 't > (input:t_option 't) (ret  )= any
    [ good -> {C_None  = input} (! ret) | bad -> {C_None  <> input} {false} any ]
    
  let rec some < 't > (input:t_option 't) (ret  (field_0:'t))= any
    [ good (field_0:'t)-> {C_Some field_0 = input} (! ret {field_0})
    | bad (field_0:'t)-> {C_Some field_0 <> input} {false} any ]
    
end
module C01_SliceFirst
  type t
  use seq.Seq
  predicate invariant'4 (self : Seq.seq t)
  let rec invariant'4 (self:Seq.seq t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'4 self} (! return' {result}) ]
    
  predicate inv'4 (_x : Seq.seq t)
  let rec inv'4 (_x:Seq.seq t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'4 _x} (! return' {result}) ]
    
  axiom inv'4 : forall x : Seq.seq t . inv'4 x = true
  use prelude.Slice
  predicate invariant'3 (self : slice t)
  let rec invariant'3 (self:slice t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'3 self} (! return' {result}) ]
    
  predicate inv'3 (_x : slice t)
  let rec inv'3 (_x:slice t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'3 _x} (! return' {result}) ]
    
  axiom inv'3 : forall x : slice t . inv'3 x = true
  use Core_Option_Option_Type as Core_Option_Option_Type
  predicate invariant'2 (self : Core_Option_Option_Type.t_option t)
  let rec invariant'2 (self:Core_Option_Option_Type.t_option t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'2 self} (! return' {result}) ]
    
  predicate inv'2 (_x : Core_Option_Option_Type.t_option t)
  let rec inv'2 (_x:Core_Option_Option_Type.t_option t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'2 _x} (! return' {result}) ]
    
  axiom inv'2 : forall x : Core_Option_Option_Type.t_option t . inv'2 x = true
  predicate invariant'1 (self : t)
  let rec invariant'1 (self:t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'1 self} (! return' {result}) ]
    
  predicate inv'1 (_x : t)
  let rec inv'1 (_x:t) (return'  (ret:bool))= any [ return' (result:bool)-> {result = inv'1 _x} (! return' {result}) ] 
  axiom inv'1 : forall x : t . inv'1 x = true
  predicate invariant'0 (self : slice t)
  let rec invariant'0 (self:slice t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = invariant'0 self} (! return' {result}) ]
    
  predicate inv'0 (_x : slice t)
  let rec inv'0 (_x:slice t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = inv'0 _x} (! return' {result}) ]
    
  axiom inv'0 : forall x : slice t . inv'0 x = true
  use seq.Seq
  use prelude.Slice
  use prelude.UIntSize
  use prelude.UIntSize
  use prelude.Int
  constant max'0 : usize = (18446744073709551615 : usize)
  function shallow_model'1 (self : slice t) : Seq.seq t
  let rec shallow_model'1 (self:slice t) (return'  (ret:Seq.seq t))= {[@expl:precondition] [#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'3 self}
    any [ return' (result:Seq.seq t)-> {result = shallow_model'1 self} (! return' {result}) ] 
  axiom shallow_model'1_spec : forall self : slice t . ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 21 19 25] inv'3 self)
   -> ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 4 19 50] inv'4 (shallow_model'1 self))
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 42] shallow_model'1 self = Slice.id self)
  && ([#"../../../../../creusot-contracts/src/std/slice.rs" 17 14 17 41] Seq.length (shallow_model'1 self)
  <= UIntSize.to_int (max'0 : usize))
  use prelude.Borrow
  function shallow_model'0 (self : slice t) : Seq.seq t =
    [#"../../../../../creusot-contracts/src/model.rs" 90 8 90 31] shallow_model'1 self
  let rec shallow_model'0 (self:slice t) (return'  (ret:Seq.seq t))= any
    [ return' (result:Seq.seq t)-> {result = shallow_model'0 self} (! return' {result}) ]
    
  use seq.Seq
  function index_logic'0 [@inline:trivial] (self : slice t) (ix : int) : t =
    [#"../../../../../creusot-contracts/src/logic/ops.rs" 43 8 43 31] Seq.get (shallow_model'1 self) ix
  let rec index_logic'0 (self:slice t) (ix:int) (return'  (ret:t))= any
    [ return' (result:t)-> {result = index_logic'0 self ix} (! return' {result}) ]
    
  use prelude.Intrinsic
  predicate resolve'1 (self : t)
  let rec resolve'1 (self:t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = resolve'1 self} (! return' {result}) ]
    
  predicate resolve'0 (self : slice t)
  let rec resolve'0 (self:slice t) (return'  (ret:bool))= any
    [ return' (result:bool)-> {result = resolve'0 self} (! return' {result}) ]
    
  let rec len'0 (self:slice t) (return'  (ret:usize))= {[@expl:precondition] inv'0 self}
    any
    [ return' (result:usize)-> {[#"../../../../../creusot-contracts/src/std/slice.rs" 238 0 347 1] Seq.length (shallow_model'0 self)
      = UIntSize.to_int result}
      (! return' {result}) ]
    
  let rec slice_first (a:slice t) (return'  (ret:Core_Option_Option_Type.t_option t))= {[#"../01.rs" 20 22 20 23] inv'0 a}
    (! bb0
    [ bb0 = s0 [ s0 = len'0 {a} (fun (_ret':usize) ->  [ &_4 <- _ret' ] s1) | s1 = bb1 ] 
    | bb1 = s0
      [ s0 = UIntSize.gt {_4} {[#"../01.rs" 21 17 21 18] (0 : usize)}
          (fun (_ret':bool) ->  [ &_3 <- _ret' ] (any [ any_ (_any:usize)-> (!  [ &_4 <- _any ] s1) ] ))
      | s1 = any [ br0 -> {_3 = false} (! bb4) | br1 -> {_3} (! bb2) ]  ]
      
    | bb2 = s0
      [ s0 =  [ &_8 <- [#"../01.rs" 22 16 22 17] (0 : usize) ] s1
      | s1 =  [ &_9 <- Slice.length a ] s2
      | s2 = UIntSize.lt {_8} {_9} (fun (_ret':bool) ->  [ &_10 <- _ret' ] s3)
      | s3 = {[@expl:index in bounds] [#"../01.rs" 22 14 22 18] _10} s4
      | s4 = bb3 ]
      
    | bb3 = s0
      [ s0 = Slice.get <t> {a} {_8} (fun (r'0:t) ->  [ &_7 <- r'0 ] s1)
      | s1 = {[@expl:type invariant] inv'0 a} s2
      | s2 = -{resolve'0 a}- s3
      | s3 = {[@expl:type invariant] inv'1 _7} s4
      | s4 = -{resolve'1 _7}- s5
      | s5 =  [ &_0 <- Core_Option_Option_Type.C_Some _7 : Core_Option_Option_Type.t_option t ] s6
      | s6 = bb5 ]
      
    | bb4 = s0
      [ s0 = {[@expl:type invariant] inv'0 a} s1
      | s1 = -{resolve'0 a}- s2
      | s2 =  [ &_0 <- Core_Option_Option_Type.C_None : Core_Option_Option_Type.t_option t ] s3
      | s3 = bb5 ]
      
    | bb5 = s0 [ s0 = return' {_0} ]  ]
    )
    [ & _0 : Core_Option_Option_Type.t_option t = any_l () : Core_Option_Option_Type.t_option t
    | & a : slice t = a
    | & _3 : bool = any_l () : bool
    | & _4 : usize = any_l () : usize
    | & _7 : t = any_l () : t
    | & _8 : usize = any_l () : usize
    | & _9 : usize = any_l () : usize
    | & _10 : bool = any_l () : bool ]
    
    [ return' (result:Core_Option_Option_Type.t_option t)-> {[@expl:postcondition] [#"../01.rs" 20 34 20 44] inv'2 result}
      {[@expl:postcondition] [#"../01.rs" 16 10 19 1] match result with
        | Core_Option_Option_Type.C_Some v -> v = index_logic'0 a 0
        | Core_Option_Option_Type.C_None -> Seq.length (shallow_model'0 a) = 0
        end}
      (! return' {result}) ]
    
end
