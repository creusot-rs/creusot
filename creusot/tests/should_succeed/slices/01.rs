extern crate creusot_contracts;

use creusot_contracts::std::*;
use creusot_contracts::*;

// fn slice_arg(a: &[u32]) -> u32 {
//     // a[10]
//     0
// }

// #[requires(10 < (@a).len() && (@a).len() <= 100)]
// fn index_slice(a: &[u32]) -> u32 {
//     a[10]
// }

// #[requires((@a).len() === 5)]
// #[ensures(@((@^a)[2]) === 3)]
// fn index_mut_slice(a: &mut [u32]) {
//     a[2] = 3;
// }

// #[ensures(match result {
//     Some(v) => *v === (@a)[0],
//     None => (@a).len() === 0
// })]
fn slice_first<T>(a: &[T]) -> Option<&T> {
    if a.len() > 0 {
        Some(&a[0])
    } else {
        None
    }
}
