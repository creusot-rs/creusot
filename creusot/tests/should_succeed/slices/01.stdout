module Type
  use Ref
  use mach.int.Int
  use prelude.Int8
  use prelude.Int16
  use mach.int.Int32
  use mach.int.Int64
  use prelude.UInt8
  use prelude.UInt16
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double
  use prelude.Prelude
  type core_option_option 't = 
    | Core_Option_Option_None
    | Core_Option_Option_Some 't
    
end
module CreusotContracts_Logic_Model_Model_ModelTy
  type self   
  type modelTy   
end
module CreusotContracts_Logic_Model_Model_Model_Interface
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Model_Model
  type self   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = self
  function model (self : self) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0_ModelTy
  type t   
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  type modelTy  = 
    ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0_Model_Interface
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  function model (self : t) : ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl0_Model
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy0 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model_Interface as Model0 with type self = t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  function model (self : t) : ModelTy0.modelTy = 
    Model0.model self
end
module CreusotContracts_Logic_Model_Impl0
  type t   
  use prelude.Prelude
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy2 with type self = t
  clone CreusotContracts_Logic_Model_Model_Model as Model2 with type self = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Impl0_Model as Model0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy,
  function Model0.model = Model2.model
  clone CreusotContracts_Logic_Model_Impl0_ModelTy as ModelTy0 with type t = t, type ModelTy0.modelTy = ModelTy2.modelTy
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = t, type ModelTy0.modelTy = ModelTy0.modelTy,
  function model = Model0.model
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = t, type modelTy = ModelTy0.modelTy
end
module CreusotContracts_Logic_Model_Impl2_ModelTy
  type t   
  use seq.Seq
  type modelTy  = 
    Seq.seq t
end
module Core_Slice_Impl0_Len_Interface
  type t   
  use seq.Seq
  use mach.int.UInt64
  use prelude.Prelude
  use mach.int.Int
  clone CreusotContracts_Logic_Model_Impl2_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = seq t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val len [@cfg:stackify] (self : seq t) : usize
    ensures { Seq.length (Model0.model self) = UInt64.to_int result }
    
end
module Core_Slice_Impl0_Len
  type t   
  use seq.Seq
  use mach.int.UInt64
  use prelude.Prelude
  use mach.int.Int
  clone CreusotContracts_Logic_Model_Impl2_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model_Interface as Model0 with type t = seq t,
  type ModelTy0.modelTy = ModelTy0.modelTy
  val len [@cfg:stackify] (self : seq t) : usize
    ensures { Seq.length (Model0.model self) = UInt64.to_int result }
    
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve_Interface
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Resolve_Resolve_Resolve
  type self   
  predicate resolve (self : self)
end
module CreusotContracts_Logic_Model_Impl2_Model_Interface
  type t   
  use prelude.Prelude
  use seq.Seq
  function model (self : seq t) : Seq.seq t
end
module CreusotContracts_Logic_Model_Impl2_Model
  type t   
  use prelude.Prelude
  use seq.Seq
  function model (self : seq t) : Seq.seq t = 
    Prelude.id self
end
module CreusotContracts_Logic_Model_Impl2
  type t   
  use prelude.Prelude
  use seq.Seq
  clone CreusotContracts_Logic_Model_Impl2_Model as Model0 with type t = t
  clone CreusotContracts_Logic_Model_Impl2_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Model_Model as Model1 with type self = seq t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function model = Model0.model
  clone CreusotContracts_Logic_Model_Model_ModelTy as ModelTy1 with type self = seq t, type modelTy = ModelTy0.modelTy
end
module C01_SliceFirst_Interface
  type t   
  use prelude.Prelude
  use seq.Seq
  use Type
  val slice_first [@cfg:stackify] (a : seq t) : Type.core_option_option t
end
module C01_SliceFirst
  type t   
  use prelude.Prelude
  use seq.Seq
  use Type
  use mach.int.Int
  use mach.int.UInt64
  clone CreusotContracts_Logic_Model_Impl2_Model as Model1 with type t = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve2 with type self = t
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve1 with type self = usize
  clone CreusotContracts_Logic_Resolve_Resolve_Resolve as Resolve0 with type self = seq t
  clone CreusotContracts_Logic_Model_Impl2_ModelTy as ModelTy0 with type t = t
  clone CreusotContracts_Logic_Model_Impl0_Model as Model0 with type t = seq t,
  type ModelTy0.modelTy = ModelTy0.modelTy, function Model0.model = Model1.model
  clone Core_Slice_Impl0_Len_Interface as Len0 with type t = t, function Model0.model = Model0.model
  let rec cfg slice_first [@cfg:stackify] (a : seq t) : Type.core_option_option t = 
  var _0 : Type.core_option_option t;
  var a_1 : seq t;
  var _2 : bool;
  var _3 : usize;
  var _4 : seq t;
  var _5 : t;
  var _6 : t;
  var _7 : usize;
  var _8 : usize;
  var _9 : bool;
  {
    a_1 <- a;
    goto BB0
  }
  BB0 {
    _4 <- a_1;
    _3 <- Len0.len _4;
    goto BB1
  }
  BB1 {
    _2 <- _3 > (0 : usize);
    switch (_2)
      | False -> goto BB4
      | _ -> goto BB2
      end
  }
  BB2 {
    _7 <- (0 : usize);
    _8 <- UInt64.of_int (Seq.length a_1);
    _9 <- _7 < _8;
    assert { _9 };
    goto BB3
  }
  BB3 {
    _6 <- Seq.get a_1 (UInt64.to_int _7);
    assume { Resolve0.resolve a_1 };
    assume { Resolve1.resolve _7 };
    _5 <- _6;
    assume { Resolve2.resolve _6 };
    _0 <- Type.Core_Option_Option_Some _5;
    goto BB5
  }
  BB4 {
    assume { Resolve0.resolve a_1 };
    _0 <- Type.Core_Option_Option_None;
    goto BB5
  }
  BB5 {
    return _0
  }
  
end
